File: vador/src/main/java/org/revcloud/vador/specs/factory/SpecFactory.java
/*******************************************************************************
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 ******************************************************************************/

package org.revcloud.vador.specs.factory;

import org.revcloud.vador.specs.specs.Spec1;
import org.revcloud.vador.specs.specs.Spec2;
import org.revcloud.vador.specs.specs.Spec3;
import org.revcloud.vador.specs.specs.Spec4;
import org.revcloud.vador.specs.specs.Spec5;

public final class SpecFactory<ValidatableT, FailureT> {

  @SuppressWarnings({"java:S100", "java:S1452"})
  public <GivenT> Spec1.Spec1Builder<ValidatableT, FailureT, GivenT, ?, ?> _1() {
    return Spec1.check();
  }

  @SuppressWarnings({"java:S100", "java:S1452"})
  public <WhenT, ThenT> Spec2.Spec2Builder<ValidatableT, FailureT, WhenT, ThenT, ?, ?> _2() {
    return Spec2.check();
  }

  @SuppressWarnings({"java:S100", "java:S1452"})
  public <WhenT, Then1T, Then2T>
      Spec3.Spec3Builder<ValidatableT, FailureT, WhenT, Then1T, Then2T, ?, ?> _3() {
    return Spec3.check();
  }

  @SuppressWarnings({"java:S100", "java:S1452"})
  public Spec4.Spec4Builder<ValidatableT, FailureT, ?, ?> _4() {
    return Spec4.check();
  }

  public Spec5.Spec5Builder<ValidatableT, FailureT, ?, ?> _5() {
    return Spec5.check();
  }
}

File: vador/src/main/kotlin/com/salesforce/vador/specs/factory/SpecFactory.kt
/**
 * ****************************************************************************
 * Copyright (c) 2022, salesforce.com, inc. All rights reserved. SPDX-License-Identifier:
 * BSD-3-Clause For full license text, see the LICENSE file in the repo root or
 * https://opensource.org/licenses/BSD-3-Clause
 * ****************************************************************************
 */
package com.salesforce.vador.specs.factory

import com.salesforce.vador.specs.specs.Spec1
import com.salesforce.vador.specs.specs.Spec2
import com.salesforce.vador.specs.specs.Spec3
import com.salesforce.vador.specs.specs.Spec4
import com.salesforce.vador.specs.specs.Spec5

class SpecFactory<ValidatableT, FailureT> {

  @Suppress("FunctionName")
  fun <GivenT> _1(): Spec1.Builder<ValidatableT, FailureT, GivenT> = Spec1.check()

  @Suppress("FunctionName")
  fun <WhenT, ThenT> _2(): Spec2.Builder<ValidatableT, FailureT, WhenT, ThenT> = Spec2.check()

  @Suppress("FunctionName")
  fun <WhenT, Then1T, Then2T> _3(): Spec3.Builder<ValidatableT, FailureT, WhenT, Then1T, Then2T> =
    Spec3.check()

  @Suppress("FunctionName") fun _4(): Spec4.Builder<ValidatableT, FailureT> = Spec4.check()

  @Suppress("FunctionName") fun _5(): Spec5.Builder<ValidatableT, FailureT> = Spec5.check()
}
